### PR TITLE
Fix OpenSSL Implementation of ValidateCertificateChain().

### PR DESCRIPTION
**Pull request by emargolis**
#### Problem
In the OpenSSL Implementation of ValidateCertificateChain() function the
intermediate certificate is loaded as a trusted certificate, which opens
door to various security attacks.

#### Change overview
Updated implementation: now loading intermediate certificate as
untrusted certificate part of cert chain.

#### Testing
Added test case that would identify this issue in the previous implementation.

all existing tests

### CODE REVIEW

**Review by bzbarsky-apple — APPROVED** [review]
Nice catch!

**Review by Damian-Nordic — APPROVED** [review]

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 1]
```diff
@@ -1574,6 +1578,7 @@ CHIP_ERROR ValidateCertificateChain(const uint8_t * rootCertificate, size_t root
     X509_free(x509LeafCertificate);
     X509_free(x509CACertificate);
     X509_free(x509RootCertificate);
+    sk_X509_pop_free(chain, X509_free);
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:1581]: Is this safe if early return done before push?

**Comment by andy31415** [comment]
@emargolis memory checker seems to complain:

```
'#3:','Test P256 Keygen                                                          ','PASSED'
[1651828999.641221][30164:30164] CHIP:CR:  ssl err  asn1 encoding routines ASN1_item_verify EVP lib

debug-malloc library: dumping program, fatal error
   Error: free space has been overwritten (err 67)
Aborted (core dumped)
```

File: src/crypto/CHIPCryptoPALOpenSSL.cpp (hunk 1 above)
Comment (emargolis) [reply src/crypto/CHIPCryptoPALOpenSSL.cpp:1581]: Yes, it should be safe

**Pull request merged**